File: src/algorithms/fixedWindow.ts
import { memoryStore } from "../stores/memoryStore";
import {redis} from "../stores/redisStore";
import {FixedWindowRecord, RateLimitResult, StoreType } from "../types";

export async function fixedWindow(
    callerId: string,
     limit: number,
      windowMs: number,
      store: StoreType
    ): Promise<RateLimitResult>{

        const now = Date.now();

        let record: FixedWindowRecord | undefined;

        if(store === "memory"){
            record = memoryStore.get(callerId);
        
        }else{
            const data = await redis.get(`fixed:${callerId}`);
            if(data){
                record = JSON.parse(data);
            }
        }

        if(!record){
            record ={
                count: 0,
                windowStart: now
            };
        }
if(now - record.windowStart >= windowMs){
    record.count = 0;
    record.windowStart = now;
}
    
  record.count++
   
  if(store === "memory"){
    memoryStore.set(callerId, record);
  }else{
    await redis.set(
        `fixed:${callerId}`,
        JSON.stringify(record)
    );
  }
    return{
        allowed: record.count <= limit,
        remaining: Math.max(limit - record.count, 0),
        retryAfter: Math.ceil(
            (windowMs -(now - record.windowStart))/1000
        )
    }
}

File: src/lua/fixedWindow.lua
local key = KEYS[1]

local limit = tonumber(ARGV[1])
local windowMs = tonumber(ARGV[2])
local now = tonumber(ARGV[3])

local count = tonumber(redis.call("HGET", key, "count"))
local windowStart = tonumber(redis.call("HGET", key, "windowStart"))

if not count then
    count = 0
    windowStart = now
end

if (now - windowStart) >= windowMs then
    count = 0
    windowStart = now
end

count = count + 1

redis.call("HSET", key,
    "count", count,
    "windowStart", windowStart
)

redis.call("PEXPIRE", key, windowMs)

local allowed = 0

if count <= limit then
    allowed = 1
end

local remaining = math.max(limit - count, 0)

local retryAfter = math.ceil(
    (windowMs - (now - windowStart)) / 1000
)

return {
    allowed,
    remaining,
    retryAfter
}